src/Stack/Dot.hs

```diff
@@ -255,37 +255,57 @@ createDepLoader :: HasEnvConfig env
                     Map FlagName Bool -> [Text] -> RIO env (Set PackageName, DotPayload))
                 -> PackageName
                 -> RIO env (Set PackageName, DotPayload)
-createDepLoader sourceMap installed globalDumpMap globalIdMap loadPackageDeps pkgName =
-  if not (pkgName `Set.member` wiredInPackages)
-      then case Map.lookup pkgName (smProject sourceMap) of
-          Just pp -> do
-            pkg <- loadCommonPackage (ppCommon pp)
-            pure (packageAllDeps pkg, payloadFromLocal pkg)
-          Nothing ->
-            case Map.lookup pkgName (smDeps sourceMap) of
-              Just DepPackage{dpLocation=PLMutable dir} -> do
-                pp <- mkProjectPackage YesPrintWarnings dir False
-                pkg <- loadCommonPackage (ppCommon pp)
-                pure (packageAllDeps pkg, payloadFromLocal pkg)
-              Just dp@DepPackage{dpLocation=PLImmutable loc} -> do
-                let common = dpCommon dp
-                gpd <- liftIO $ cpGPD common
-                let PackageIdentifier name version = PD.package $ PD.packageDescription gpd
-                    flags = cpFlags common
-                    ghcOptions = cpGhcOptions common
-                assert (pkgName == name) (loadPackageDeps pkgName version loc flags ghcOptions)
-              Nothing ->
-                pure (Set.empty, payloadFromInstalled (Map.lookup pkgName installed))
-      -- For wired-in-packages, use information from ghc-pkg (see #3084)
-      else case Map.lookup pkgName globalDumpMap of
-          Nothing -> error ("Invariant violated: Expected to find wired-in-package " ++ packageNameString pkgName ++ " in global DB")
-          Just dp -> pure (Set.fromList deps, payloadFromDump dp)
-            where
-              deps = map (\depId -> maybe (error ("Invariant violated: Expected to find " ++ ghcPkgIdString depId ++ " in global DB"))
-                                          Stack.Prelude.pkgName
-                                          (Map.lookup depId globalIdMap))
-                         (dpDepends dp)
+createDepLoader sourceMap installed globalDumpMap globalIdMap loadPackageDeps pkgName = do
+  case projectPackageDeps <|> dependencyDeps <|> globalDeps of
+    Just x -> x
+    Nothing ->
+      if isWiredInPackage then
+        let errText = "Invariant violated: Expected to find wired-in-package "
```

Comment:
Any reason to have this `errText` as a separate variable?

Reply:
BTW I'm not sure that we need special handling for wired-in packages at all though I'm OK with leave it as it is

Reply:
I was just using `errText` to pretty up the code a bit, it looked a little ugly having all those string additions on one line. But if wired-in and global are basically the same thing, then I could just remove all of this. I was just worried about removing the whole invariant violated error. In the original code, wired-in-packages not existing in the `globalDumpMap` caused an error, but non-wired-in global packages were truncated. Should global packages missing from the dump map always cause error? Or should they just return an empty dependency set? This is the main thing im concerned with right now, knowing this will make it a little easier to remove `createDepLoader` as well. Thanks for all your help with this issue!

Reply:
After looking at it again, it seems like getting rid of `createDepLoader` will be a bit harder than I thought. It doesn't necessarily need to exist, but It's helping to keep the code organized as it kind of unifies all the dependency loaders and makes the arguments they need available to them. So if it's decided that we should remove that function, I'll leave that for another PR.